send/ldap_it4i

```diff
@@ -0,0 +1,166 @@
+#!/usr/bin/perl
+use strict;
+use warnings;
+no if $] >= 5.017011, warnings => 'experimental::smartmatch';
+use Net::LDAPS;
+use Net::LDAP::Entry;
+use Net::LDAP::Message;
+use Net::LDAP::LDIF;
+use ScriptLock;
+use ADConnector;
+
+sub process_update;
+sub process_archive;
+
+# Constants and counters
+my $counter_add = 0;
+my $counter_update = 0;
+my $counter_archive = 0;
+my $counter_fail = 0;
+
+# Define service
+my $service_name = "ldap_it4i";
+
+# GEN folder location
+my $facility_name = $ARGV[0];
+chomp($facility_name);
+my $service_files_base_dir="../gen/spool";
+my $service_files_dir="$service_files_base_dir/$facility_name/$service_name";
+my $service_file = "$service_files_dir/$service_name.ldif";
+
+# BASE DN
+open my $file, '<', "$service_files_dir/baseDN";
+my $base_dn = <$file>;
+chomp($base_dn);
+close $file;
+
+# propagation destination
+my $namespace = $ARGV[1];
+chomp($namespace);
+
+# create service lock
+my $lock = ScriptLock->new($facility_name . "_" . $service_name . "_" . $namespace);
+($lock->lock() == 1) or die "Unable to get lock, service propagation was already running.";
+
+# init configuration
+my @conf = init_config($namespace);
+# we cannot use a function to resolve PDC from DNS since the LDAP instance is not an AD, therefore, we use the raw value
+my $ldap_location = $conf[0];
+my $ldap = ldap_connect($ldap_location);
+my $filter = '(objectClass=person)';
+
+# connect
+ldap_bind($ldap, $conf[1], $conf[2]);
+
+# load all data
+my @users = load_perun($service_file);
+my @previousUsers = load_ad($ldap, $base_dn, $filter, ['uid','givenName','sn','fullName','displayName','mail','userPassword','state','sshPublicKey']);
+
+my %previousUsersMap = ();
+foreach my $previousUser (@previousUsers) {
+	my $login = $previousUser->get_value('uid');
+	$previousUsersMap{ $login } = $previousUser;
+}
+
+my %usersMap = ();
+foreach my $user (@users) {
+	my $login = $user->get_value('uid');
+	$usersMap{ $login } = $user;
+}
+
+# process data
+process_update();
+process_archive();
+
+# disconnect
+ldap_unbind($ldap);
+
+# log results
+ldap_log($service_name, "Added: " . $counter_add . " entries.");
+ldap_log($service_name, "Updated: " . $counter_update. " entries.");
+ldap_log($service_name, "Archived: " . $counter_archive. " entries.");
+ldap_log($service_name, "Failed: " . $counter_fail. " entries.");
+
+$lock->unlock();
+
+if ($counter_fail > 0) { die "Failed to process: " . $counter_fail . " entries.\nSee log at: ~/perun-engine/send/logs/$service_name.log";}
+
+# private methods
+sub process_update() {
+
+	foreach my $user (@users) {
+
+		if (exists $previousUsersMap{$user->get_value('uid')}) {
+
+			my $previousUser = $previousUsersMap{$user->get_value('uid')};
+
+			# attrs without cn since it's part of DN to be updated
```

Comment:
We are using `uid` as primary attribute used in DN, so you can add `cn` between attributes for update and fix this comment.

Please update `load_perun()` and `load_ad()` functions in `ADConnector.pm`. There are checks for required attributes of entry, without which entries are discarded from returned result. We must add `uid` between conditions in order to get entries for this service to be processed. Thank you.

Reply:
Ok, added, thanks.